eras/shelley/impl/src/Cardano/Ledger/Shelley/LedgerState/DPState.hs

```diff
@@ -235,19 +268,166 @@ instance Default (DState c) where
 ptrsMap :: DState c -> Map Ptr (Credential 'Staking c)
-ptrsMap (DState (UnifiedMap _ ptrmap) _ _ _) = ptrmap
+ptrsMap (DState (UnifiedMap _ ptrmap) _ _ _ _) = ptrmap
+
+-- ==========================================================
+-- Functions that handle Deposits for stake credetials and key hashes.
+
+-- | One only pays a deposit on the initial key registration. If the key has been
+--   de-registered it should have been removed from the map. If it hasn't been
+--   de-registered, then it has no effect on the Deposits. In places where this function
+--   is called, there should be an explicit check that the credential is not in the map.
+payKeyDeposit ::
+  HasField "_keyDeposit" (PParams era) Coin =>
+  Credential 'Staking (EraCrypto era) ->
+  PParams era ->
+  DState (EraCrypto era) ->
+  DState (EraCrypto era)
+payKeyDeposit cred pp dstate = dstate {dsDeposits = newStake}
+  where
+    stake = dsDeposits dstate
+    newStake = case Map.lookup cred stake of
+      Nothing -> Map.insert cred (getField @"_keyDeposit" pp) stake
+      Just _ -> stake
+
+refundKeyDeposit :: Credential 'Staking c -> DState c -> (Coin, DState c)
+refundKeyDeposit cred dstate =
+  (coin, dstate {dsDeposits = newStake})
+  where
+    stake = dsDeposits dstate
+    (coin, newStake) = case Map.lookup cred stake of
+      Just c -> (c, Map.delete cred stake)
+      Nothing -> (mempty, stake)
+
+-- | One only pays a deposit on the initial pool registration. So return the
+--   the Deposits unchanged if the keyhash already exists. There are legal
+--   situations where a pool may be registered multiple times.
+payPoolDeposit ::
+  HasField "_poolDeposit" (PParams era) Coin =>
+  KeyHash 'StakePool (EraCrypto era) ->
+  PParams era ->
+  PState (EraCrypto era) ->
+  PState (EraCrypto era)
+payPoolDeposit keyhash pp pstate = pstate {psDeposits = newpool}
+  where
+    pool = psDeposits pstate
+    newpool = case Map.lookup keyhash pool of
+      Nothing -> Map.insert keyhash (getField @"_poolDeposit" pp) pool
+      Just _ -> pool -- Should it be overwritten with the current  (getField @"_poolDeposit" pp) ?
+      -- things are simpler if it is not.
+
+refundPoolDeposit :: KeyHash 'StakePool c -> PState c -> (Coin, PState c)
+refundPoolDeposit keyhash pstate = (coin, pstate {psDeposits = newpool})
+  where
+    pool = psDeposits pstate
+    (coin, newpool) = case Map.lookup keyhash pool of
+      Just c -> (c, Map.delete keyhash pool)
+      Nothing -> (mempty, pool)
+
+-- | Calculate total possible refunds in the system. There is an invariant that
+--   this should be the same as the utxosDeposited field of the UTxOState. Note that
+--   this does not depend upon the current values of the Key and Pool deposits of the PParams.
+obligationDPState :: DPState era -> Coin
+obligationDPState (DPState DState {dsDeposits = keys} PState {psDeposits = stakePools}) =
+  foldl' (<>) (Coin 0) keys <> foldl' (<>) (Coin 0) stakePools
```

Comment:
some day soon, dsDeposits and psDeposits are going to hold CompactCoin, so unless we make CompactCoin an instance of Val (which I don't think we can because of the method 'invert') am going to leave this as it is.